### PR TITLE
[Auth] Implement authentication

### LINKED ISSUE #192: Authentication package
## Feature Description

An authentication package for both JS & Python.

## Problem You're Trying to Solve

Help the developers by providing a secure by default library for them to use

### PR DESCRIPTION
**Pull request by MNThomson**
# Description

Implement the auth package for use in both python and JS

Closes #192

## Checklist
- [x] The code includes tests if relevant
- [ ] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

File: apps/backend/lib/src/auth.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import { Request, Response, NextFunction } from "express";
+import * as jwt from "jsonwebtoken";
+import { Algorithm } from "jsonwebtoken";
```
Comment (Scott-Kenning) [inline apps/backend/lib/src/auth.ts:3]: Can we clean up this import? Preferabbly just import whats needed from jsonwebtoken

File: apps/backend/lib/src/auth.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  userId: number;
+}
+
+export interface AuthenticatedRequest extends Request {
+  user: UserToken;
+}
+
+const SIGNOPTIONS: jwt.SignOptions = {
+  algorithm: "RS256" as Algorithm,
+  expiresIn: "14 days",
```
Comment (Scott-Kenning) [inline apps/backend/lib/src/auth.ts:15]: I think we agreed on a shorter expiry, but this is probably much nicer for dev

File: apps/backend/lib/src/auth.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  // @ts-ignore
+  let decoded: jwt.JwtPayload & UserToken;
+
+  const authCookie = req.cookies["authorization"];
+  try {
+    decoded = jwt.verify(authCookie, JWT_PUBLIC_KEY, { algorithms: ["RS256"] });
+  } catch (error) {
+    console.error(error);
+    res.status(401);
```
Comment (Scott-Kenning) [inline apps/backend/lib/src/auth.ts:38]: can we have a json message attached to this with a descriptive error

Comment (Scott-Kenning) [reply apps/backend/lib/src/auth.ts:38]: same goes for the other returns

Comment (MNThomson) [reply apps/backend/lib/src/auth.ts:38]: Api spec says no response body

File: apps/backend/lib/src/auth.ts (hunk 2 above)
Comment (MNThomson) [reply apps/backend/lib/src/auth.ts:15]: I don't recall discussing an expiration time, but I have no issues changing this in the future

**Comment by MNThomson** [comment]
This PR is feature complete so I'll be passing it off to @UVicMartletplace/backend to get the tests passing (all auth tests are green, but the initial backend tests written are extremely flaky and don't take auth into consideration)

**Review by TobySMurray — APPROVED** [review]
LGTM, nice job

**Review by Scott-Kenning — APPROVED** [review]

**Pull request merged**